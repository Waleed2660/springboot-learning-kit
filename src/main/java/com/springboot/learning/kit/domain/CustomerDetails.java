package com.springboot.learning.kit.domain;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@Builder
@Table(name = "customer_details")
@NoArgsConstructor
@AllArgsConstructor
public class CustomerDetails {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;
    private String email;
    private String phone;
}
