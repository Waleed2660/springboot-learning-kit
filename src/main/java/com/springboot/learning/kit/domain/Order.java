package com.springboot.learning.kit.domain;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Getter
@Builder
@Table(name = "orders")
@NoArgsConstructor
@AllArgsConstructor
public class Order {
    @Id
    @Column(name = "uuid")
    private Long uuid;

    @Column(name = "order_type")
    private OrderType orderType;

    @JoinColumn(name = "customer_details_id")
    private Long customerDetailsId;

    @JoinColumn(name = "customer_address_id")
    private Long customerAddressId;

    @Column(name = "total_amount")
    private BigDecimal totalAmount;

    @Column(name = "currency")
    private String currency;

    @Column(name = "order_created")
    private LocalDateTime orderCreated;
}
