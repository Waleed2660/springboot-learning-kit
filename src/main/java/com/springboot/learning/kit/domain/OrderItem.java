package com.springboot.learning.kit.domain;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;

@Entity
@Getter
@Builder
@Table(name = "order_item")
@NoArgsConstructor
@AllArgsConstructor
public class OrderItem {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "order_id")
    private Long orderId;

    @Column(name = "product_id")
    private Long productId;

    @Column(name = "quantity")
    private int quantity;

    @Column(name = "status")
    private String status;

    @Column(name = "price_per_unit")
    private BigDecimal pricePerUnit;
}
