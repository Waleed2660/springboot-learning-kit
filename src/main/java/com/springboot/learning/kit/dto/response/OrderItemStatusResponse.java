package com.springboot.learning.kit.dto.response;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;
import lombok.Data;

@Data
@Builder
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class OrderItemStatusResponse {
    private Long productId;
    private Integer quantity;
    private String status;
}
