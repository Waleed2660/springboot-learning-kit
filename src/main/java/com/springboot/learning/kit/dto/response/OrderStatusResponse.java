package com.springboot.learning.kit.dto.response;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.Builder;
import lombok.Data;

import java.time.LocalDateTime;
import java.util.List;

@Data
@Builder
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class OrderStatusResponse {
    private Long orderId;
    private String orderStatus;
    private String orderType;
    private LocalDateTime orderCreated;
    private List<OrderItemStatusResponse> items;
}
