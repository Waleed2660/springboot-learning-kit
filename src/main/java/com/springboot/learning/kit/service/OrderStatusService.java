package com.springboot.learning.kit.service;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class OrderStatusService {

    public void saveOrderStatus() {

    }
}
