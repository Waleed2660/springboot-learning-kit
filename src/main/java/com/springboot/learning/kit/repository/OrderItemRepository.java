package com.springboot.learning.kit.repository;

import com.springboot.learning.kit.domain.OrderItem;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface OrderItemRepository extends JpaRepository<OrderItem, Long> {

    /**
     * Find all order items belonging to the given orderId
     * @param orderId the order's ID
     * @return list of order items
     */
    List<OrderItem> findByOrderId(Long orderId);

}
