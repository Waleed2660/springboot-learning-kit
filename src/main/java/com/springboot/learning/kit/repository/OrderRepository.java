package com.springboot.learning.kit.repository;

import com.springboot.learning.kit.domain.Order;
import com.springboot.learning.kit.domain.OrderType;
import java.util.List;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface OrderRepository extends JpaRepository<Order, Long> {

    // Custom query methods can be defined here
    // For example, find orders by status or type
    default List<Order> findByType(OrderType type) {
        return null;
    }
}
