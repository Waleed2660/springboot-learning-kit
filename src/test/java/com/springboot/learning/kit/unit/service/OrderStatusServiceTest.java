package com.springboot.learning.kit.unit.service;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.springboot.learning.kit.domain.Order;
import com.springboot.learning.kit.domain.OrderItem;
import com.springboot.learning.kit.domain.OrderType;
import com.springboot.learning.kit.dto.response.OrderItemStatusResponse;
import com.springboot.learning.kit.dto.response.OrderStatusResponse;
import com.springboot.learning.kit.repository.OrderItemRepository;
import com.springboot.learning.kit.repository.OrderRepository;
import com.springboot.learning.kit.service.OrderStatusService;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

import static org.mockito.Mockito.when;

@ExtendWith({MockitoExtension.class, SnapshotExtension.class})
class OrderStatusServiceTest {

    @Mock
    private OrderRepository orderRepository;

    @Mock
    private OrderItemRepository orderItemRepository;

    @InjectMocks
    private OrderStatusService orderStatusService;

    private Expect expect;

    @Test
    void verifyOrderStatusResponse() {
        // Arrange
        long orderUUID = 1L;

        // Mock the repository method that fetches the order
        when(orderRepository.findById(orderUUID)).thenReturn(Optional.of(createOrder(orderUUID)));

        // Mock the repository method that fetches the order items
        when(orderItemRepository.findByOrderId(orderUUID)).thenReturn(createOrderItems(orderUUID));

        // Act
        OrderStatusResponse status = orderStatusService.getOrderStatus(orderUUID);

        // Assert
        expect.serializer("json").toMatchSnapshot(status);
    }

    /**
     * Creates an order with the given order ID.
     * @param orderId
     * @return Order
     */
    private @NotNull Order createOrder(long orderId) {
        return Order.builder()
                .uuid(orderId)
                .orderType(OrderType.ONLINE)
                .customerDetailsId(101L)
                .customerAddressId(201L)
                .totalAmount(BigDecimal.valueOf(99.99))
                .currency("GBP")
                .orderCreated(java.time.LocalDateTime.now())
                .build();
    }

    /**
     * Creates a list of order items for the given order ID.
     * @param orderId
     * @return List<OrderItem>
     */
    private @NotNull List<OrderItem> createOrderItems(long orderId) {
        return List.of(
                OrderItem.builder()
                        .orderId(orderId)
                        .productId(101L)
                        .quantity(2)
                        .status("CONFIRMED")
                        .pricePerUnit(BigDecimal.valueOf(49.99))
                        .build(),
                OrderItem.builder()
                        .orderId(orderId)
                        .productId(102L)
                        .quantity(1)
                        .status("PENDING")
                        .pricePerUnit(BigDecimal.valueOf(10.50))
                        .build()
        );
    }
}
