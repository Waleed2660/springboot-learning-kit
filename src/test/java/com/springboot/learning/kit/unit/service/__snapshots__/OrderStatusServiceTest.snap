com.springboot.learning.kit.unit.service.OrderStatusServiceTest.verifyOrderStatusResponse=[
  {
    "orderId": 1,
    "orderType": "ONLINE",
    "items": [
      {
        "productId": 101,
        "quantity": 2,
        "status": "CONFIRMED"
      },
      {
        "productId": 102,
        "quantity": 1,
        "status": "PENDING"
      }
    ]
  }
]